iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,572 @@
+{
+  locationtype l = this->itemMap.find(key);
+  if (l.isValid()) {
+    // found
+    return make_ptr((RefCountCacheItem<C> *)l.m_value->item.get());
+  } else {
+    return Ptr<RefCountCacheItem<C>>(NULL);
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::del(uint64_t key)
+{
+  this->itemMap.remove(key);
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::clear()
+{
+  // Clear the in memory hashmap
+  // TODO: delete all items (not sure if clear calls delete on all items)
+  this->itemMap.clear();
+  this->items = 0;
+  this->size = 0;
+}
+
+template <class C>
+int
+RefCountCachePartition<C>::numItems()
+{
+  return this->itemMap.count();
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::copy(std::vector<Ptr<RefCountCacheItemBase>> *items)
+{
+  for (RefCountCachePartition<C>::iteratortype i = this->itemMap.begin(); i != this->itemMap.end(); ++i) {
+    items->push_back(make_ptr((RefCountCacheItemBase *)i->item.get()));
+  }
+}
+
+// TODO: pass an iterator or something? Seems not good to pass a pointer to a member
+template <class C>
+TSHashTable<RefCountCacheHashing> *
+RefCountCachePartition<C>::getMap()
+{
+  return &this->itemMap;
+}
+
+// The implementation of this class must be in here, as this is used by template
+// classes, and c++ rocks that way
+// The header for the cache, this is used to check if the serialized cache is compatible
+class RefCountCacheHeader
+{
+public:
+  unsigned int magic;
+  VersionNumber version;
+
+  int numPartitions;
+  int partitionMaxSize;
+
+  RefCountCacheHeader(int numParts, int partSize)
+    : magic(REFCOUNTCACHE_MAGIC_NUMBER), numPartitions(numParts), partitionMaxSize(partSize)
+  {
+    this->version.ink_major = REFCOUNTCACHE_MAJOR_VERSION;
+    this->version.ink_minor = REFCOUNTCACHE_MINOR_VERSION;
+  };
+
+  bool
+  operator==(const RefCountCacheHeader other) const
+  {
+    return (this->magic == other.magic && this->version.ink_major == other.version.ink_major &&
+            this->version.ink_minor == other.version.ink_minor && this->numPartitions == other.numPartitions &&
+            this->partitionMaxSize == other.partitionMaxSize);
+  }
+
+  bool
+  compatible(RefCountCacheHeader *other)
+  {
+    //      printf("magic %d %d\n", this->magic, other->magic);
+    //      printf("version.ink_major %d %d\n", this->version.ink_major, other->version.ink_major);
+    //      printf("numPartitions %d %d\n", this->numPartitions, other->numPartitions);
+    //      printf("partitionMaxSize %d %d\n", this->partitionMaxSize, other->partitionMaxSize);
+
+    return (this->magic == other->magic && this->version.ink_major == other->version.ink_major &&
+            this->numPartitions == other->numPartitions && this->partitionMaxSize == other->partitionMaxSize);
+  };
+};
+
+// Base for syncing purposes
+// Since the actual RefCountCache is templated, we don't want to require
+// importing the template type for all uses (such as syncing)
+class RefCountCacheBase
```

Comment:
Although `Base` may be an ugly name-- it is quite consistent with the rest of the codebase (from a quick grep). I've updated the PR changing the various things to be const that should be :)